### PR TITLE
Surface point density 2× (65k→131k): denser surface sampling on SOTA stack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Doubling the surface point density from 65,536 to 131,072 points per sample will improve surface pressure and wall shear predictions by providing the model with a denser geometric representation of the car surface. The yi-track recorded that increasing width from 256d to 512d improved volume_pressure (15.21→14.37%) suggesting capacity changes help the model capture fine-grained surface structure. Here we test a *data resolution* lever orthogonal to architecture — more surface points means the model sees finer boundary layer features, turbulent separation zones, and surface curvature variations that are aliased at 65k.

**Key motivation:**
- Surface pressure (4.44% vs AB-UPT 3.82%) and wall shear (8.00% vs AB-UPT 7.29%) remain above reference — denser sampling directly helps surface-level predictions
- tau_y/z are the largest gaps; tau directional patterns live on the surface and depend on local surface geometry resolution
- No cost to model capacity; pure data resolution lever
- Untested on tay track — this is a first test

**Note on surface point count choice:** `--train-surface-points 131072` is a 2× increase from current 65536 baseline. A value of 196608 (3×) is also feasible if memory allows, but start with 2× (131072) to establish the trend clearly. If the pod has 96GB VRAM and batch-size=4, 131072 should be within budget.

## Instructions

Run the full SOTA stack (PR #387: STRING-sep, QK-norm, feat16 RFF, Lion lr=1e-4, EMA=0.999) with **surface points doubled from 65536 to 131072**:

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 131072 --eval-surface-points 131072 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 \
  --wandb-group nezuko-surf-pts-r24 \
  --wandb-name nezuko/surf-pts-131k-sota-stack
```

**Important:** If the 131072 surface points run causes OOM (out-of-memory) errors, fall back to `--train-surface-points 98304 --eval-surface-points 98304` (1.5× = 98304) and note the OOM in your PR comment.

**Gate at EP5:** val_abupt must be ≤ 9.5% by EP5, or slope must be −0.3+ pp/1k-steps.

**Report at completion:**
1. W&B run ID and group
2. Val trajectory (every epoch)
3. Best val_abupt and the epoch it occurred
4. Test metrics from best-val checkpoint: test_abupt, surface_pressure, wall_shear, volume_pressure, tau_x, tau_y, tau_z
5. Per-epoch training time (to compare vs SOTA 22 min/epoch)
6. GPU memory usage peak (nvidia-smi or logged)

## Baseline (PR #387 alphonse — current SOTA)

| Metric | SOTA PR #387 | AB-UPT |
|---|---:|---:|
| val_abupt (best EP11) | **7.3816%** | — |
| test_abupt | **8.5936%** | — |
| test surface_pressure | 4.4377% | 3.82% |
| test wall_shear | 7.9989% | 7.29% |
| test volume_pressure | 12.1885% | 6.08% |
| test tau_x | 6.9622% | 5.35% |
| test tau_y | 9.1058% | 3.65% |
| test tau_z | 10.2736% | 3.63% |
| Train surface pts | 65536 | — |
| ~22 min/epoch | — | — |

W&B run: `wj6mn6ve` · group `alphonse-rff-sweep`

**Target to beat: val_abupt < 7.3816% and/or test_abupt < 8.5936%**

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Status update noted:
- mlp-ratio-r22 finished with val_abupt=7.5137% (+0.132pp regression vs SOTA 7.3816%) — wider MLP didn't help enough
- surf-pts-r24 run `e4gz48nf` launched at EP0.79 (no val yet) — too early to evaluate
- smoke run `5u5r08yh` crashed — noted

The surf-pts=32768 hypothesis is worth testing since 24K (r24) is already running. Please also confirm the r24 run is using the same hyperparameters as the r22 run (same lr, wd, batch size) so we have a clean comparison. Let me know when r24 reaches EP5 with gate results.

**Comment by morganmcg1** [comment]
STUDENT nezuko: Status update

**r24 hyperparameter confirmation:** r24 (surf-pts-131k, run `e4gz48nf`) uses the **same lr/wd/batch-size/optimizer/architecture** as r22 (mlp-ratio-8). The only intentional differences vs r22:
- `train/eval-surface-points`: **131072** (vs r22's 65536) — the experimental lever
- r22 had `mlp-ratio-8` (which I'm not applying here)

Common config (matches SOTA #387 stack): `--optimizer lion --lr 1e-4 --weight-decay 5e-4 --batch-size 4 --no-compile-model --validation-every 1 --train-volume-points 65536 --eval-volume-points 65536 --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 --pos-encoding-mode string_separable --use-qk-norm --rff-num-features 16`.

**EP1 result (run `e4gz48nf`):**
- val_primary/abupt_axis_mean_rel_l2_pct = **52.4663%**
- epoch_time = **2194 s** (~36.6 min/epoch, vs SOTA ~22 min/epoch — 1.66× slower as expected with 2× surface points)
- peak GPU = **85.4 GB** (well under 96 GB budget; no OOM)

**Comparison to SOTA #387 (`wj6mn6ve`) at EP1:** 52.79% — our run is essentially identical to SOTA at EP1 (−0.33pp). This is a positive sign — we're on track and the doubled point count hasn't perturbed early dynamics.

**ETA on EP5:** training started 07:01 UTC, Epoch 2/50 is at 56% as of 07:59 UTC. EP5 should finish around 10:05 UTC; will report gate result then.

**Timeout note:** SENPAI_TIMEOUT_MINUTES=360. At 36.6 min/epoch, we'll fit ~9-10 epochs in the timeout window. SOTA reached its best at EP11 — we may run out of time before matching that exact best-val. The hypothesis hinges on whether 131k points improves *per-epoch* convergence enough to compensate for the slower epochs.

**Comment by morganmcg1** [comment]
ADVISOR: Nezuko — status received. Your r24 run `e4gz48nf` looks healthy at EP1 (52.47% vs SOTA EP1 ~52.79%, essentially identical). The config confirmation is clean — the only difference from the SOTA stack is 131k surface points.

**EP5 gate threshold: val_abupt ≤ 9.5%** (stricter than usual given the ~22% per-epoch penalty vs SOTA's convergence pace). At 36.6 min/epoch you have a ~9-10 epoch budget in the 360 min timeout window.

**Key question at EP5:** Is the per-epoch val metric lower than the SOTA at the same epoch? SOTA (run `r5rw40rn`) at EP5 was likely around 9.0-9.5%. If 131k pts is converging faster per epoch than 65k pts, it may compensate for having fewer epochs. Report EP5 SOTA comparison if you have it.

**If EP5 gate passes:** Let it run to natural timeout — the val at EP9-10 is what matters, not whether it reaches EP11. Run the best checkpoint eval regardless.

**If EP5 gate fails (>9.5%):** Kill and post the data — the hypothesis that denser points accelerate per-epoch convergence isn't panning out and we should close this rather than burn the remaining timeout.

Also: I asked you to confirm whether r24 uses the same hyperparams as r22 (mlp-ratio-8). Your response addressed the SOTA stack comparison but note that r22 had `mlp-ratio-8` which is *not* part of the SOTA stack. So the comparison is r24 (131k pts, standard stack) vs SOTA `r5rw40rn` (65k pts, standard stack) — that's the right control. Good.

Report EP5 results when ready.

**Comment by morganmcg1** [comment]
STUDENT nezuko: EP5 gate update — **GATE PASSES** (9.13% ≤ 9.5%)

## EP5 result (run `e4gz48nf`)
- **val_abupt = 9.1316%** ✅ (gate threshold ≤9.5%)
- val_surface_pressure = 5.9138%
- val_wall_shear = 10.137%
- val_volume_pressure = 5.9044%
- val_tau_x/y/z = 8.659% / 11.935% / 13.246%
- best_checkpoint flag: still updating each epoch (EP5 latest improvement)

## Per-epoch comparison vs SOTA #387 (`wj6mn6ve`, 65k surface pts, same stack)

| EP | SOTA 65k val_abupt | Ours 131k val_abupt | Δ |
|---|---:|---:|---:|
| EP1 | 52.7945% | 52.4663% | −0.33pp |
| EP2 | 29.2198% | 30.6161% | +1.40pp |
| EP3 | 12.8939% | 13.9127% | +1.02pp |
| EP4 | 9.9467% | 10.2846% | +0.34pp |
| EP5 | **8.8573%** | **9.1316%** | **+0.27pp** |

**Key observation:** 131k surface points lags SOTA at every epoch after EP1 (small but consistent gap). The deficit is shrinking (1.40 → 1.02 → 0.34 → 0.27pp), so 131k is converging *closer* to SOTA pace per epoch from EP3 onward, but it has not closed the gap by EP5.

## Wall-clock cost
- epoch_time = **~2195 s/epoch (36.6 min)** vs SOTA `wj6mn6ve` 1478 s (24.6 min) → **1.49× slower** per epoch
- Peak GPU memory: **~88.7 GB / 97.9 GB** (no OOM, 9 GB headroom)
- On a compute-equivalent basis at 5×SOTA epochs (~123 min), we are at val=9.13% while SOTA at the equivalent ~5 epochs of 65k cost (~5 epochs of 1478s ≈ 123 min) is at **val=8.86%**. Compute-normalized, denser points is currently losing.

## Timeline
- Started 07:01 UTC, currently mid-EP6 at 10:18 UTC
- SENPAI_TIMEOUT_MINUTES=360 → finish by ~13:01 UTC
- At 36.6 min/epoch we'll fit **~9-10 epochs** within the timeout (vs SOTA's 11)

## Plan
Per advisor: gate passes → let it run to natural timeout, then run best-checkpoint eval. The val at EP9-10 is what matters. Will re-check around EP9.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #506 because Run  finished at 282.7min. Best val: 7.9581% (EP7), test: 9.071%. Doesn't beat current SOTA 7.1792%/8.497%. The 2x surface points (131k) hypothesis doesn't help — slower per-epoch iteration means fewer epochs fit in the budget (~7 epochs vs 11 for SOTA). The model sees fewer full training passes, which outweighs any benefit from denser surface sampling. Lesson: surface resolution is already saturated at 65k pts — the bottleneck is depth of training, not surface sampling density. Closing as dead end. Thank you nezuko for clean results..

**Comment by morganmcg1** [comment]
STUDENT nezuko:

## Results

**Hypothesis:** Doubling surface points 65k → 131k improves surface predictions.
**Outcome:** ❌ **Negative.** On a fixed wall-clock budget, denser sampling regresses on every primary metric. Per-epoch, it reaches parity with SOTA from EP6 onward, but the 1.49× per-epoch slowdown means we cannot reach the convergence epochs (EP9–11) where SOTA bottoms out.

### Headline metrics (best-val checkpoint, EP8)

| Metric | SOTA #387 (65k, EP11) | Ours 131k (EP8) | Δ vs SOTA | AB-UPT ref |
|---|---:|---:|---:|---:|
| **val_abupt (best)** | **7.3816%** | **7.9581%** | **+0.58pp ❌** | — |
| **test_abupt** | **8.5936%** | **9.0707%** | **+0.48pp ❌** | — |
| test surface_pressure | 4.4377% | 4.8326% | +0.39pp ❌ | 3.82% |
| test wall_shear | 7.9989% | 8.6895% | +0.69pp ❌ | 7.29% |
| test volume_pressure | 12.1885% | 11.9264% | −0.26pp ✅ | 6.08% |
| test tau_x | 6.9622% | 7.5619% | +0.60pp ❌ | 5.35% |
| test tau_y | 9.1058% | 10.0256% | +0.92pp ❌ | 3.65% |
| test tau_z | 10.2736% | 11.0068% | +0.73pp ❌ | 3.63% |

### Val trajectory per epoch (vs SOTA `wj6mn6ve` 65k, same stack)

| EP | SOTA val_abupt | Ours val_abupt | Δ | Notes |
|---|---:|---:|---:|---|
| 1 | 52.7945% | 52.4663% | −0.33pp | nearly identical start |
| 2 | 29.2198% | 30.6161% | +1.40pp | |
| 3 | 12.8939% | 13.9127% | +1.02pp | |
| 4 | 9.9467% | 10.2846% | +0.34pp | |
| 5 | 8.8573% | 9.1316% | +0.27pp | gate ✅ (≤9.5%) |
| 6 | 8.7305% | 8.7066% | **−0.02pp** | first parity epoch |
| 7 | 8.0005% | 8.0573% | +0.06pp | parity continues |
| 8 (partial) | 7.9883% | **7.9581%** | **−0.03pp** | best val; partial epoch (~25% of EP8 before timeout) |
| 9 | 7.5917% | — | — | (SOTA continues) |
| 10 | 7.4538% | — | — | |
| 11 | **7.3816%** | — | — | SOTA best |

**Per-epoch parity from EP6 onward** — the denser-points run converges at the same rate as SOTA after warmup. The val gap closes from +1.40pp at EP2 to ≤0.06pp from EP6. **However**, we ran out of time at EP8 (partial) while SOTA reached EP11. The 1.49× wall-clock penalty more than offsets the per-epoch parity.

### Wall-clock cost

| | SOTA 65k (`wj6mn6ve`) | Ours 131k (`e4gz48nf`) | Ratio |
|---|---:|---:|---:|
| Per-epoch time | 1478 s (24.6 min) | 2194 s (36.6 min) | 1.49× |
| Train budget | 270 min | 270 min | — |
| Epochs in budget | 11 | 7.25 | 0.66× |
| Best epoch | 11 | 8 (partial) | — |

### Compute / memory
- Peak GPU memory: **93.0 GB / 97.9 GB (~90.6%)** — high but no OOM, ~5 GB headroom
- All 8 GPUs DDP, batch-size=4, 8×4=32 effective
- Volume points unchanged at 65k (only surface lever moved)

### Command used
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 131072 --eval-surface-points 131072 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 \
  --wandb-group nezuko-surf-pts-r24 \
  --wandb-name nezuko/surf-pts-131k-sota-stack
```

### W&B
- Run ID: **`e4gz48nf`** (rank 0; ranks 1-7: `pm4ct4ml`, `cx2ijx4p`, `3n5me355`, `h96qkjaj`, `gjcxv47s`, `ye6ks8q5`, `n9zb2grs`)
- Group: `nezuko-surf-pts-r24`
- Best epoch: 8 (EMA checkpoint)
- Total train minutes: 270.68

### What happened — analysis

The hypothesis assumed denser surface sampling would either **(a) accelerate per-epoch convergence** or **(b) lift the achievable val/test ceiling**. The data shows neither:

1. **No per-epoch acceleration in early/mid training.** EP2-5 are *worse* by +0.27 to +1.40pp. The denser cloud appears to add stochasticity that early training fights through, not exploits.
2. **Per-epoch parity from EP6 onward.** Once the model has learned the gross flow structure, the additional points neither help nor hurt at fixed epoch budget. This is the most surprising sub-result — it means denser sampling is essentially a wash per-epoch, not a gain.
3. **Wall-clock loses badly.** 1.49× slower per epoch + same train budget = 7.25 epochs vs SOTA's 11. SOTA squeezes ~3.5 extra epochs of refinement (EP9-11 where val drops 7.99% → 7.38%), and we miss them all.
4. **Test/val gap holds as expected.** Test ~+1pp above val for both runs (8.59 vs 7.38 for SOTA; 9.07 vs 7.96 for us). No structural change in generalization.
5. **Volume pressure surprise (−0.26pp).** We didn't change volume sampling, so this is likely noise from the different best epoch (EP8 vs EP11 — at our EP8 with our setup, vol_p mae=9.61 is already healthier than SOTA's at EP11). Not a meaningful lever finding.

**Why it failed:** the prior on this hypothesis ("more surface resolution captures finer boundary-layer features") implicitly assumed the model is *bottlenecked* by surface point density at 65k. The data says it isn't — surface predictions plateau at the same per-epoch rate regardless of 65k vs 131k. The actual bottlenecks for surface metrics (4.44% sp, 8.00% ws, ~10% tau) appear to live in **architecture/training duration**, not data resolution.

### Suggested follow-ups

Ranked by likely-impact / cost:

1. **Don't pursue 196k (3×).** The 131k result already shows no per-epoch gain and a 1.5× cost penalty; 196k would be ~2× cost penalty for the same null result. Skip.
2. **Surface points = 32-49k (downscale lever).** Symmetric counter-test: if 131k is no better per-epoch, is 65k itself overkill? Halving to 32k or 49k might give the same val with ~0.7× wall time, leaving room for more epochs. This is the more promising direction.
3. **Mixed sampling — dense at high-curvature regions, sparse elsewhere.** Curvature-weighted importance sampling on the surface mesh would put points where they matter (separation lines, tau_y/z spikes) at lower total point cost. Larger engineering lift but addresses the actual bottleneck (tau directional metrics).
4. **Longer train budget (compute-budget lever).** If we believe per-epoch parity holds, running 131k for 13-15 epochs (with extended SENPAI_TIMEOUT_MINUTES=540) could test whether denser sampling matches/beats SOTA when given matching epoch counts. But at 1.5× cost this is mostly a compute-budget question, not a data-resolution one.
5. **Higher batch-size offset.** With 5GB headroom we could try batch-size=2 + 131k pts to roughly match wall-clock per epoch and isolate the per-epoch effect more cleanly. But this conflates batch-size and pts changes — risky.

**Bottom line:** Doubling surface points is a wash per-epoch and a loss per-wall-clock. Recommend closing this direction and pursuing tau-component fixes via architectural levers (e.g., dedicated tau heads, curvature-aware features) rather than data resolution.
